bluetooth: host: avdtp: Fix command and response format error

CC0918 opened the pull request:
There are three aspects of changes involved:
1.  The format of Service Category should be checked  in set configuration cmd and error code and Service Category should be detailed in response.
2. There is no  INT Stream Endpoint ID in re_config cmd,.
3. Fail ACP SEID should be added to start and suspend rej rsp.

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+		}
+	}
+
+	return 0;
```
use `BT_AVDTP_SUCCESS`

CC0918 replied:
Done

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -451,17 +582,25 @@ static void avdtp_process_configuration_cmd(struct bt_avdtp *session, struct net
 		} else if (buf->len >= 1U) {
-			uint8_t int_seid;
-
+			uint8_t int_seid = 0;
```
Add one blank line after declaring variables.

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -451,17 +582,25 @@ static void avdtp_process_configuration_cmd(struct bt_avdtp *session, struct net
+				int_seid = net_buf_pull_u8(buf) >> 2;
+			}
+			net_buf_simple_save(&buf->b, &state);
+			err = bt_avdtp_check_service_category(buf, &service_category, reconfig);
```
suggest to use another variable (like: `err_code`) because `err` is `int`, but `bt_avdtp_check_service_category` return `uint8_t`.

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -451,17 +582,25 @@ static void avdtp_process_configuration_cmd(struct bt_avdtp *session, struct net
 			if (!reconfig) {
-				err = session->ops->set_configuration_ind(session, sep, int_seid,
-									  buf, &avdtp_err_code);
+				/* int seid not in reconfig cmd*/
```
I see the "set configuration" cmd has ACP SEID and INT SEID, "reconfigure" cmd has only ACP SEID. But there is no code to pull_u8 for the ACP SEID, it seems a bug, please help to check.

CC0918 replied:
The code to pull_u8 for the ACP SEID in func avdtp_get_cmd_sep().

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+	uint8_t data;
+	uint8_t losc;
+
+	if (!buf) {
```
I think it is not needed because there are many `buf` process codes before this function and the process codes don't check whether buf is NULL.

CC0918 replied:
Done

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+
+	while (buf->len) {
+		data = net_buf_pull_u8(buf);
+		*service_category = data;
```
add blank line after/before code block.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+			break;
+
+		case BT_AVDTP_SERVICE_MEDIA_CODEC:
+			if (buf->len < 3U) {
```
```suggestion
			if (buf->len < szieof(losc)) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+			}
+
+			losc = net_buf_pull_u8(buf);
+			if (losc != 3U || losc > buf->len) {
```
I think we should consider adding a structure for the payload and use the `sizeof` to check the buf length instead of using hard code `3`.

CC0918 replied:
add structure and macro in avdtp_internel.h

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+			losc = net_buf_pull_u8(buf);
+			if (losc != 0U) {
+				if (buf->len < losc) {
+					return BT_AVDTP_BAD_ACP_SEID;
```
Ditto

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+
+			losc = net_buf_pull_u8(buf);
+			if (losc < 2U || losc > buf->len) {
+				return BT_AVDTP_BAD_ACP_SEID;
```
Ditto

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+
+		case BT_AVDTP_SERVICE_MEDIA_CODEC:
+			if (buf->len < 3U) {
+				return BT_AVDTP_BAD_ACP_SEID;
```
Ditto

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -451,17 +582,25 @@ static void avdtp_process_configuration_cmd(struct bt_avdtp *session, struct net
+						session, sep, int_seid, buf, &avdtp_err_code);
+				} else {
+					err = session->ops->re_configuration_ind(
+						session, sep, int_seid, buf, &avdtp_err_code);
```
Done

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+		return BT_AVDTP_BAD_ACP_SEID;
+	}
+
+	while (buf->len) {
```
Done

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+		*service_category = data;
+		switch (data) {
+		case BT_AVDTP_SERVICE_MEDIA_TRANSPORT:
+			if (buf->len < 1U) {
```
Done

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,141 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+			break;
+
+		case BT_AVDTP_SERVICE_MEDIA_RECOVERY:
+			if (buf->len < 4U) {
```
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,143 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+		case BT_AVDTP_SERVICE_HEADER_COMPRESSION:
+		case BT_AVDTP_SERVICE_REPORTING:
+		case BT_AVDTP_SERVICE_DELAY_REPORTING:
+			if (buf->len < sizeof(losc)) {
+				return BT_AVDTP_BAD_LENGTH;
+			}
+
+			losc = net_buf_pull_u8(buf);
+			if (losc != 0U) {
+				if (buf->len < losc) {
+					return BT_AVDTP_BAD_LENGTH;
+				}
+				net_buf_pull_mem(buf, losc);
+			}
+
+			if (reconfig) {
+				return BT_AVDTP_INVALID_CAPABILITIES;
+			}
+			break;
+
+		case BT_AVDTP_SERVICE_MEDIA_CODEC:
+			if (buf->len < sizeof(struct bt_avdtp_media_codec_capabilities)) {
+				return BT_AVDTP_BAD_LENGTH;
+			}
+
+			losc = net_buf_pull_u8(buf);
+			if (losc < MEDIA_CODEC_CAPABILITIES_MIN_LOSC || losc > buf->len) {
+				return BT_AVDTP_BAD_LENGTH;
+			}
+
+			net_buf_pull_mem(buf, losc);
+			break;
+
+		default:
+			return BT_AVDTP_BAD_SERV_CATEGORY;
+		}
+	}
+
+	return BT_AVDTP_SUCCESS;
+}
```
Suggested reference code,
```
struct bt_avdtp_service_category_handler {
	uint16_t min_len;
	uint16_t max_len;
	bool reconfig;
	uint8_t error;
	uint8_t (*handler)(struct net_buf *buf, uint8_t *service_category);
};

static struct bt_avdtp_service_category_handler = category_handler[] {
...
};

struct bt_avdtp_generic_service_cap {
	uint8_t service_category;
	uint8_t losc;
} __packed;

uint8_t bt_avdtp_check_service_category(struct net_buf *buf, uint8_t *service_category, bool reconfig)
{
	struct bt_avdtp_generic_service_cap *hdr;
	uint8_t err;

	if (buf->len == 0U) {
		LOG_DBG("Error: buf not valid");
		return BT_AVDTP_BAD_LENGTH;
	}

	while (buf->len > 0U) {
		if (buf->len < sizeof(*hdr)) {
			LOG_DBG("Error: buf not valid");
			return BT_AVDTP_BAD_LENGTH;
		}

		hdr = net_buf_simple_pull_mem(buf, sizeof(*hdr));

		if (hdr->service_category != 0 && hdr->service_category < ARRAY_SIZE(category_handler)) {
			*service_category = hdr->service_category;

			if (hdr->losc > buf->len || hdr->losc > category_handler[hdr->service_category - 1].max_len || hdr->losc < category_handler[hdr->service_category - 1].min_len) {
				return category_handler[hdr->service_category - 1].error;
			}

			if (!category_handler[hdr->service_category - 1].reconfig && reconfig) {
				return BT_AVDTP_INVALID_CAPABILITIES;
			}

			if (category_handler[hdr->service_category - 1].handler != NULL) {
				err = category_handler[hdr->service_category - 1].handler(buf);
				if (err != BT_AVDTP_SUCCESS) {
					return err;
				}
			}
			net_buf_simple_pull(buf, hdr->losc);
		} else {
			return BT_AVDTP_BAD_SERV_CATEGORY;
		}
	}

	return BT_AVDTP_SUCCESS;
}
```

CC0918 replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+	 sizeof(struct bt_avdtp_multiplexing_capabilities),
+	 0, BT_AVDTP_BAD_MULTIPLEXING_FORMAT, NULL},
+	{sizeof(struct bt_avdtp_media_codec_capabilities), UINT8_MAX, 1, BT_AVDTP_BAD_LENGTH, NULL},
+	{0, 0, 0, BT_AVDTP_BAD_LENGTH, NULL},
```
It is not needed.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+	}
+	return BT_AVDTP_SUCCESS;
+}
+static struct bt_avdtp_service_category_handle category_handler[]  = {
```
Empty line is missing.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+	uint8_t max_len;
+	bool reconfig_support;
+	uint8_t err_code;
+	uint8_t (*handle)(struct net_buf *buf, uint8_t *len);
```
```suggestion
	uint8_t (*handler)(struct net_buf *buf, uint8_t *len);
```

CC0918 replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+	uint8_t recovery_type = net_buf_pull_u8(buf);
+	*len -= sizeof(recovery_type);
+
+	if (recovery_type != BT_AVDTP_RECOVERY_TYPE_FORBIDDEN &&
+		recovery_type != BT_ADVTP_RECOVERY_TYPE_RFC2733) {
```
```suggestion
	if (recovery_type != BT_AVDTP_RECOVERY_TYPE_FORBIDDEN &&
	    recovery_type != BT_ADVTP_RECOVERY_TYPE_RFC2733) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+			}
+
+			if (service_handle.handle != NULL) {
+				err = service_handle.handle(buf, &(hdr->losc));
```
The parameter `len` is not needed. My suggestion is,
```suggestion
				struct net_buf_simple_state state;

				net_buf_simple_save(&buf->b, &state);
				err = service_handle.handle(buf, &(hdr->losc));
				net_buf_simple_restore(&buf->b, &state);
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
 	}
 }
 
+struct bt_avdtp_service_category_handle {
```
```suggestion
struct bt_avdtp_service_category_handler {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+{
+	struct bt_avdtp_generic_service_cap *hdr;
+	uint8_t err;
+	struct bt_avdtp_service_category_handle service_handle;
```
```suggestion
	struct bt_avdtp_service_category_handle handler;
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,102 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
+
+		if (hdr->service_category != 0 &&
+		    hdr->service_category < ARRAY_SIZE(category_handler)) {
+			service_handle = category_handler[hdr->service_category];
```
```suggestion
			handler = &category_handler[hdr->service_category];
```

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -438,6 +568,9 @@ static void avdtp_process_configuration_cmd(struct bt_avdtp *session, struct net
 		err = -ENOTSUP;
+	} else if (!reconfig && sep->sep_info.inuse == 1) {
+		avdtp_err_code = BT_AVDTP_SEP_IN_USE;
+		err = -ENOTSUP;
```
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp_internal.h:
```diff
@@ -173,6 +173,46 @@ struct bt_avdtp_ctrl_params {
+	uint8_t cp_type_spec_value[0];
+} __packed;
+
+struct bt_avdtp_header_compression_capabilities {
+	uint8_t backch : 1;
+	uint8_t media : 1;
+	uint8_t recovery : 1;
+} __packed;
```
The bit field of structure has the byte-order issue.
```issue
struct bt_avdtp_header_compression_capabilities {
#if defined(CONFIG_LITTLE_ENDIAN)
	uint8_t reserved : 5;
	uint8_t recovery : 1;
	uint8_t media : 1;
	uint8_t backch : 1;
#else
	uint8_t backch : 1;
	uint8_t media : 1;
	uint8_t recovery : 1;
	uint8_t reserved : 5;
#endif /* defined(CONFIG_LITTLE_ENDIAN) */
} __packed;
```

CC0918 committed:
bluetooth: host: avdtp: Fix command and response format error

There are three aspects of changes involved:
1.  The format of Service Category should be checked  in set
configuration cmd and error code and Service Category
should be detailed in response.
2. There is no  INT Stream Endpoint ID in re_config cmd,.
3. Fail ACP SEID should be added to start and suspend rej rsp.

Signed-off-by: Cheng Chang <cheng.chang@nxp.com>

CC0918 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1519,7 +1630,8 @@ int bt_avdtp_parse_capability_codec(struct net_buf *buf, uint8_t *codec_type,
 static int avdtp_process_configure_command(struct bt_avdtp *session, uint8_t cmd,
-					   struct bt_avdtp_set_configuration_params *param)
+					   struct bt_avdtp_set_configuration_params *param,
+					   bool reconfig)
```
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -421,13 +421,108 @@ static void avdtp_get_capabilities_rsp(struct bt_avdtp *session, struct net_buf
 	struct net_buf *rsp_buf;
 	uint8_t avdtp_err_code = 0;
+	struct net_buf_simple_state state;
+	uint8_t service_category = 0;
```
According to 
```
Contains the value of the first Service Category to fail. In case no
Service Category applies due to the nature of the error, the Service
Category field shall be set to 0x0.
```,  the 0 is a correct value.
